cli/azd/pkg/project/service_target_containerapp.go

```diff
@@ -27,175 +28,277 @@ import (
+type containerAppPackageResult struct {
+	ImageTag    string
+	LoginServer string
 }
 
-func (at *containerAppTarget) Deploy(
-	ctx context.Context,
-	azdCtx *azdcontext.AzdContext,
-	path string,
-	progress chan<- string,
-) (ServiceDeploymentResult, error) {
-	// If the infra module has not been specified default to a module with the same name as the service.
-	if strings.TrimSpace(at.config.Infra.Module) == "" {
-		at.config.Infra.Module = at.config.Module
-	}
-	if strings.TrimSpace(at.config.Infra.Module) == "" {
-		at.config.Infra.Module = at.config.Name
-	}
-
-	// Login to container registry.
-	loginServer, has := at.env.Values[environment.ContainerRegistryEndpointEnvVarName]
-	if !has {
-		return ServiceDeploymentResult{}, fmt.Errorf(
-			"could not determine container registry endpoint, ensure %s is set as an output of your infrastructure",
-			environment.ContainerRegistryEndpointEnvVarName,
-		)
-	}
-
-	log.Printf("logging into registry %s", loginServer)
-
-	progress <- "Logging into container registry"
-	if err := at.containerRegistryService.LoginAcr(ctx, at.env.GetSubscriptionId(), loginServer); err != nil {
-		return ServiceDeploymentResult{}, fmt.Errorf("logging into registry '%s': %w", loginServer, err)
-	}
-
-	imageTag, err := at.generateImageTag()
-	if err != nil {
-		return ServiceDeploymentResult{}, fmt.Errorf("generating image tag: %w", err)
-	}
-
-	fullTag := fmt.Sprintf(
-		"%s/%s",
-		loginServer,
-		imageTag,
-	)
-
-	// Tag image.
-	log.Printf("tagging image %s as %s", path, fullTag)
-	progress <- "Tagging image"
-	if err := at.docker.Tag(ctx, at.config.Path(), path, fullTag); err != nil {
-		return ServiceDeploymentResult{}, fmt.Errorf("tagging image: %w", err)
-	}
-
-	log.Printf("pushing %s to registry", fullTag)
-
-	// Push image.
-	progress <- "Pushing container image"
-	if err := at.docker.Push(ctx, at.config.Path(), fullTag); err != nil {
-		return ServiceDeploymentResult{}, fmt.Errorf("pushing image: %w", err)
+// NewContainerAppTarget creates the container app service target.
+//
+// The target resource can be partially filled with only ResourceGroupName, since container apps
+// can be provisioned during deployment.
+func NewContainerAppTarget(
+	env *environment.Environment,
+	containerRegistryService azcli.ContainerRegistryService,
+	azCli azcli.AzCli,
+	docker docker.Docker,
+	console input.Console,
+	commandRunner exec.CommandRunner,
+	accountManager account.Manager,
+	serviceManager ServiceManager,
+	resourceManager ResourceManager,
+) ServiceTarget {
+	return &containerAppTarget{
+		env:                      env,
+		accountManager:           accountManager,
+		serviceManager:           serviceManager,
+		resourceManager:          resourceManager,
+		cli:                      azCli,
+		containerRegistryService: containerRegistryService,
+		docker:                   docker,
+		console:                  console,
+		commandRunner:            commandRunner,
+		clock:                    clock.New(),
```

Comment:
Should this have been injected?

Reply:
This is how this was but I can update it quick.